returnn/datasets/basic.py

```diff
@@ -639,11 +657,19 @@ def _apply_partition_epoch(cls, seq_index, partition_epoch, epoch):
         assert sum(partition_sizes) == num_seqs and len(partition_sizes) == partition_epoch
         partitions = functools.reduce(lambda a, x: a + [a[-1] + x], partition_sizes, [0])  # cumulative sum
         assert len(partitions) == partition_epoch + 1
-        seq_index = seq_index[partitions[current_partition] : partitions[current_partition + 1]]
-        assert len(seq_index) == partition_sizes[current_partition]
-
+        seq_index = seq_index[shard_index:][::num_shards][
```

Comment:
Why do you do the logic twice, and also in a different way?

Just do the logic once, but instead of `partition_epoch`, use `partition_epoch * num_shards`. And then the index is `current_partition * num_shards + shard_index`.

There are almost no code changes here in this function (except maybe rename some of the vars).

Reply:
If you split the data like you said you get consecutive seq IDs for the different sub workers. This is ok for the workers themselves, but makes it difficult (or does it?) in MultiProcDataset to look up which worker to query for which seq, because `seq_idx % num_workers` no longer gives the correct worker index.

I think `seq_idx // num_seqs_per_worker` would then give the correct worker index again, but if you don't know `num_seqs_per_worker` you're out of luck. Moreover, if you iterate the through the segments front to back, `worker_idx = seq_idx // num_seqs_per_worker` means you first query worker 1 for all the segments it has, then worker 2, and so on. This means you lose the benefit from having more than one worker in the first place, because the workers do not get any time to fill their buffers in the background while other workers are queried for segments.

Currently MultiProcDS distributes the `seq_order` shards therefore also in a strided way between the workers and I think it should stay that way. Or am I wrong here somewhere?

Reply:
I still don't exactly understand where you see the problem. The test is just wrong. But you added the test?

Reply:
Oh, then ok. Yeah I copied and adapted an existing test case, and I assumed we should leave the order as-is.